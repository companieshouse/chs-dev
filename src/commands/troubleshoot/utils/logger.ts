import { simpleColouriser } from "../../../helpers/colouriser.js";

export default class TroubleshootOutputLogger {

    private readonly logger: (msg: string) => void;

    constructor (logger: (msg: string) => void) {
        this.logger = logger;
    }

    logMessage (message: string, stat: "Success" | "Error"): void {
        stat === "Success"
            ? this.logger(`${simpleColouriser("SUCCESS!", "bold-green")} ${message}`)
            : this.logger(`${simpleColouriser("FAILED!", "red")} ${message}`);

    }
}
