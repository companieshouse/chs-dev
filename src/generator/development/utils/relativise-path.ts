import { dirname, join, relative } from "path";

/**
 * Creates relative paths from the destination directory to the source for the
 * supplied paths supplied.
 * @param paths Array of strings or string representing the paths being
 *      converted to relative paths
 * @param source original location from which the path was relative to prior
 * @param destinationDirectory Location from which the new path will be
 *      relative from
 * @returns relative path from the destination directory to the paths provided
 */
export const relativiseServicePath = (
    paths: string | string[],
    source: string,
    destinationDirectory: string
) => {
    if (Array.isArray(paths)) {
        return paths.map(
            singleValue => relativiseServicePath(singleValue, source, destinationDirectory)
        );
    }

    const relativePathToSource = relative(destinationDirectory, source);

    return join(dirname(relativePathToSource), paths);
};

export default relativiseServicePath;
