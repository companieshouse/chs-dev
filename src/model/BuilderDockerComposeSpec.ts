/**
 * Represents a Docker Compose specification for a builder
 */
export interface BuilderDockerComposeSpec {
    /**
     * Name of the builder
     */
    name: string;

    /**
     * Version of the builder
     */
    version: string;

    /**
     * String representation of the docker compose spec - may contain variables
     * in '< >' which are removed when the docker compose file is created
     */
    builderSpec: string;
}

export default BuilderDockerComposeSpec;
