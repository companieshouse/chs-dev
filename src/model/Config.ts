/**
 * Represents project configuration defining project related properties for
 * customising the behaviours of chs-dev
 */
export type Config = {

    /**
     * name to refer to the project/environment
     */
    readonly projectName: string;

    /**
     * path to the project repository
     */
    readonly projectPath: string;

    /**
     * path to the chs-dev local repository
     */
    readonly chsDevPath?: string;

    /**
     * path to the chs-dev data
     */
    readonly chsDevDataDir?: string;

    /**
     * additional environment variables for running the environment. Values can
     * contain 'file://' prefixes will try to read file in and set as value in
     * its place.
     */
    readonly env: Record<string, string>;

    /**
     * when supplied defines the threshold for number of hours after which to
     * attempt ecr login
     */
    readonly performEcrLoginHoursThreshold?: number;

    /**
     * specifies the supported version of the CLI by the project
     */
    readonly versionSpecification?: string;

}

export default Config;
