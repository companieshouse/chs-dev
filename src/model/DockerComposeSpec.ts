type Secret = {
    environment: string
} | {
    file: string
}

/**
 * Represents the Docker Compose Spec file when read in from yaml
 */
export type DockerComposeSpec = {
    services: {
        [serviceName: string] : {
            labels?: string[] | Record<string, string>,
            environment?: string[] | Record<string, string>,
            image?: string,
            build?: Record<string, any>,
            env_file?: string | string[],
            depends_on?: string[] | Record<string, Record<string, any>>
            secrets?: string[] | Record<string, any>,
            ports?: string | string[]
        } & Record<string, any>
    },
    secrets?: Record<string, Secret>,
    include?: string[],
    networks?: Record<string, any>,
    volumes?: Record<string, any>
};
