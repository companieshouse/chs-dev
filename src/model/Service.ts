import { DependencyNode } from "./DependencyGraph.js";
import { OrNull } from "./OrNull.js";

/**
 * Represents a service within the development environment. A service is a
 * single deployable unit - typically container.
 */
export interface Service {
    /**
     * Meaningful name for service, unique within the environment
     */
    name: string;

    /**
     * Name of the module which the service resides
     */
    module: string;

    /**
     * Human readable description of the service
     */
    description?: OrNull<string>;

    /**
     * Path to the Docker Compose file for the service
     */
    source: string;

    /**
     * List of services names (exhaustive) that the service depends on.
     *
     * This will include all of the direct and indirect/transitive dependencies
     * the service has.
     */
    dependsOn: string[];

    /**
     * Tree of the dependencies for the service.
     *
     * This will include all of the direct and indirect/transitive dependencies
     * the service has.
     */
    dependencyTree?: DependencyNode;

    /**
     * The number of times this service is used by other services.
     *
     * This is the number of services that have this service in their `dependsOn` list and
     * is not the same as `numberOfDependencies` which counts the number of
     * services that the service depends on.
     */
    timesUsedByOtherServices?: number;
    /**
     * The number of services that depend on this service.
     *
     * This is the number of services that this service depends on either directly or
     * indirectly. This is not the same as `timesUsedByOtherServices` which counts
     * the number of time a service appears on the `dependsOn` list.
     */
    numberOfDependencies?: number;
    /**
     * Details about the code repository for the service (required for
     * development mode)
     */
    repository: OrNull<{ url: string; branch?: string | null; }>;

    /**
     * When supplied will specify a common Dockerfile to use to build the
     * service for development mode
     */
    builder: string;

    /**
     * Other pertinent information about the service which may vary
     * service to service.
     */
    metadata: Record<string, OrNull<string> | undefined>;
}

export default Service;
