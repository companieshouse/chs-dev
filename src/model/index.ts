export enum LogCoverage {
  WATCH,
  LOG,
}

export enum Prune {
  VOLUME,
  NETWORK,
  IMAGE,
  CONTAINER,
  BUILDER,
}
