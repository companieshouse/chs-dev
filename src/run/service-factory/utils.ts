import NewServiceSpec from "../../model/NewServiceSpec.js";

/**
 * Apply the value to the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param selector Jq like selector to apply the value to
 * @param newServiceSpec The newServiceSpec to apply the value to
 * @param newValue The value to apply
 * @returns The newServiceSpec with the value applied
 */
export const applySelector = (newServiceSpec: Partial<NewServiceSpec>, selector: string, newValue: any) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return {
            ...newServiceSpec,
            [nestedKeys[0]]: applySelector(
                newServiceSpecSubAttribute,
                `.${nestedKeys.slice(1).join(".")}`,
                newValue
            )
        };
    } else {
        return {
            ...newServiceSpec,
            [nestedKeys[0]]: newValue
        };
    }
};

/**
 * Lookup a value in the newServiceSpec at the selector. The selector can
 * reference a nested object using dot notation. For example, ".configuration.docker-chs-development.module"
 * but not using array notation (i.e. .configuration["docker-chs-development"]["module"]).
 * @param newServiceSpec NewServiceSpec to lookup the value in
 * @param selector Jq like selector to lookup the value at
 * @returns The value at the selector
 */
export const get = (newServiceSpec: Partial<NewServiceSpec>, selector: string) => {
    const nestedKeys = extractValidNestedKeys(selector);

    if (nestedKeys.length > 1) {
        const newServiceSpecSubAttribute = getNextLevelObject(newServiceSpec, nestedKeys[0]);

        return get(newServiceSpecSubAttribute, `.${nestedKeys.slice(1).join(".")}`);
    } else {
        return newServiceSpec[nestedKeys[0]];
    }
};

const isSelectorValid = (selector: string) => {
    return /^\.([A-Za-z0-9_-]+\.?)+$/.test(selector);
};

const extractValidNestedKeys = (selector: string) => {
    if (!isSelectorValid(selector)) {
        throw new Error("Invalid selector");
    }

    return selector.split(".").slice(1);
};

const getNextLevelObject = (newServiceSpec: Partial<NewServiceSpec>, nestedKey: string) => {
    return newServiceSpec[nestedKey] || {};
};
