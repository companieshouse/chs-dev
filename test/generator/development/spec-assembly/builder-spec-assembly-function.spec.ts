import { expect } from "@jest/globals";
import yaml from "yaml";
import builderSpecAssemblyFunction from "../../../../src/generator/development/spec-assembly/builder-spec-assembly-function";
import BuilderDockerComposeSpec from "../../../../src/model/BuilderDockerComposeSpec";
import { DockerComposeSpec } from "../../../../src/model/DockerComposeSpec";
import { services } from "../../../utils/data";
import { generateBuilderSpec, generateServiceSpec } from "../../../utils/docker-compose-spec";
import Service from "../../../../src/model/Service";
import { join } from "path";

describe("builderSpecAssemblyFunction", () => {

    const service = services[3];
    const builderLocation = "local/builders/java/v2";
    const projectPath = "/home/test/user/projects/docker-project";
    const serviceDockerComposeSpec = generateServiceSpec(service);

    let builderDockerComposeSpec: BuilderDockerComposeSpec;

    beforeEach(() => {
        builderDockerComposeSpec = {
            builderSpec: generateBuilderSpec(
                builderLocation, true
            ),
            name: "java",
            version: "v2"
        };
    });

    it("adds extra builder service to the generated spec", () => {
        const developmentDockerComposeSpec = {
            services: {
                [service.name]: {}
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            service,
            projectPath,
            serviceDockerComposeSpec,
            builderDockerComposeSpec
        });

        expect(developmentDockerComposeSpec.services).toMatchSnapshot();
    });

    it("adds dependency builder on service and can handle no dependencies present", () => {
        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [service.name]: {}
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            service,
            projectPath,
            serviceDockerComposeSpec,
            builderDockerComposeSpec
        });

        expect(developmentDockerComposeSpec.services).toMatchSnapshot();
    });

    it("adds dependency builder on service and can handle dependencies present", () => {
        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [service.name]: {
                    depends_on: {
                        "service-two": {
                            condition: "service_started"
                        }
                    }
                }
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            service,
            projectPath,
            serviceDockerComposeSpec,
            builderDockerComposeSpec
        });

        expect(developmentDockerComposeSpec.services).toMatchSnapshot();
    });

    it("applies extra properties to service", () => {
        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [service.name]: {}
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            service,
            projectPath,
            serviceDockerComposeSpec,
            builderDockerComposeSpec
        });

        expect(developmentDockerComposeSpec.services).toMatchSnapshot();
    });

    it("can handle repository builder", () => {
        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [service.name]: {}
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            service,
            projectPath,
            serviceDockerComposeSpec,
            builderDockerComposeSpec: {
                name: "repository",
                version: "v1",
                builderSpec: yaml.stringify({
                    services: {
                        "<service>": {
                            build: {
                                context: "<absolute_repository_path>"
                            }
                        }
                    }
                })
            }
        });

        expect(developmentDockerComposeSpec.services).toMatchSnapshot();
    });

    it("can handle setting repository context for repo builder", () => {
        const serviceWithContext = { ...services[7] };
        const repoContext = "tilt/";
        serviceWithContext.metadata.repoContext = repoContext;

        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [serviceWithContext.name]: {}
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            projectPath,
            serviceDockerComposeSpec: generateServiceSpec(serviceWithContext),
            service: serviceWithContext,
            builderDockerComposeSpec: {
                name: "repository",
                version: "v1",
                builderSpec: yaml.stringify({
                    services: {
                        "<service>": {
                            build: {
                                context: "<absolute_repository_path>"
                            }
                        }
                    }
                })
            }
        });

        expect(developmentDockerComposeSpec.services[serviceWithContext.name]).toMatchSnapshot();
    });

    it("can handle setting dockerfile for repo builder", () => {
        const serviceWithDockerfile: Service = { ...services[2] };
        const dockerfile = "tilt/Dockerfile";
        serviceWithDockerfile.metadata.dockerfile = dockerfile;

        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [serviceWithDockerfile.name]: {}
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            projectPath,
            serviceDockerComposeSpec: generateServiceSpec(serviceWithDockerfile),
            service: serviceWithDockerfile,
            builderDockerComposeSpec: {
                name: "repository",
                version: "v1",
                builderSpec: yaml.stringify({
                    services: {
                        "<service>": {
                            build: {
                                context: "<absolute_repository_path>"
                            }
                        }
                    }
                })
            }
        });

        expect(developmentDockerComposeSpec.services[serviceWithDockerfile.name]).toMatchSnapshot();
    });

    it("can handle setting dockerfile and context for repo builder", () => {
        const serviceWithDockerfileAndContext: Service = { ...services[3] };
        const context = "another-dir/";
        const dockerfile = "another.Dockerfile";
        serviceWithDockerfileAndContext.metadata.dockerfile = dockerfile;
        serviceWithDockerfileAndContext.metadata.repoContext = context;

        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [serviceWithDockerfileAndContext.name]: {}
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            projectPath,
            serviceDockerComposeSpec: generateServiceSpec(serviceWithDockerfileAndContext),
            service: serviceWithDockerfileAndContext,
            builderDockerComposeSpec: {
                name: "repository",
                version: "v1",
                builderSpec: yaml.stringify({
                    services: {
                        "<service>": {
                            build: {
                                context: "<absolute_repository_path>"
                            }
                        }
                    }
                })
            }
        });

        expect(developmentDockerComposeSpec.services[serviceWithDockerfileAndContext.name]).toMatchSnapshot();
    });

    it("can handle using builder and service dockerfile", () => {
        const serviceBuildServiceFromRepo: Service = { ...services[0] };
        serviceBuildServiceFromRepo.metadata.builderUseServiceDockerFile = "true";

        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [serviceBuildServiceFromRepo.name]: {}
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            projectPath,
            serviceDockerComposeSpec: generateServiceSpec(serviceBuildServiceFromRepo),
            service: serviceBuildServiceFromRepo,
            builderDockerComposeSpec
        });

        expect(developmentDockerComposeSpec.services[serviceBuildServiceFromRepo.name]).toMatchSnapshot();

        expect(developmentDockerComposeSpec.services[`${serviceBuildServiceFromRepo.name}-builder`]).toMatchSnapshot();
    });

    it("can handle using builder and service dockerfile appending custom context and dockerfile", () => {
        const serviceBuildServiceFromRepo: Service = { ...services[0] };
        serviceBuildServiceFromRepo.metadata.builderUseServiceDockerFile = "true";

        const context = "another-dir/";
        const dockerfile = "another.Dockerfile";
        serviceBuildServiceFromRepo.metadata.dockerfile = dockerfile;
        serviceBuildServiceFromRepo.metadata.repoContext = context;

        const developmentDockerComposeSpec: DockerComposeSpec = {
            services: {
                [serviceBuildServiceFromRepo.name]: {}
            }
        };

        builderSpecAssemblyFunction(developmentDockerComposeSpec, {
            projectPath,
            serviceDockerComposeSpec: generateServiceSpec(serviceBuildServiceFromRepo),
            service: serviceBuildServiceFromRepo,
            builderDockerComposeSpec
        });

        expect(developmentDockerComposeSpec.services[serviceBuildServiceFromRepo.name]).toMatchSnapshot();

        expect(developmentDockerComposeSpec.services[`${serviceBuildServiceFromRepo.name}-builder`]).toMatchSnapshot();
    });
});
