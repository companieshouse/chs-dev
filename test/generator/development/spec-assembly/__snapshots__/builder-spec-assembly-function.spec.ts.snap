// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builderSpecAssemblyFunction adds dependency builder on service and can handle dependencies present 1`] = `
{
  "service-four": {
    "build": {
      "context": "/home/test/user/projects/docker-project",
      "dockerfile": "local/builders/java/v2/Dockerfile",
    },
    "depends_on": {
      "service-four-builder": {
        "condition": "service_completed_successfully",
        "restart": true,
      },
      "service-two": {
        "condition": "service_started",
      },
    },
    "volumes": [
      "repositories/service-four/target:/opt/",
    ],
  },
  "service-four-builder": {
    "build": {
      "context": "/home/test/user/projects/docker-project",
      "dockerfile": "local/builders/java/v2/build.Dockerfile",
    },
    "volumes": [
      "repositories/service-four:/app/",
      "./out:/opt/out",
    ],
  },
}
`;

exports[`builderSpecAssemblyFunction adds dependency builder on service and can handle no dependencies present 1`] = `
{
  "service-four": {
    "build": {
      "context": "/home/test/user/projects/docker-project",
      "dockerfile": "local/builders/java/v2/Dockerfile",
    },
    "depends_on": {
      "service-four-builder": {
        "condition": "service_completed_successfully",
        "restart": true,
      },
    },
    "volumes": [
      "repositories/service-four/target:/opt/",
    ],
  },
  "service-four-builder": {
    "build": {
      "context": "/home/test/user/projects/docker-project",
      "dockerfile": "local/builders/java/v2/build.Dockerfile",
    },
    "volumes": [
      "repositories/service-four:/app/",
      "./out:/opt/out",
    ],
  },
}
`;

exports[`builderSpecAssemblyFunction adds extra builder service to the generated spec 1`] = `
{
  "service-four": {
    "build": {
      "context": "/home/test/user/projects/docker-project",
      "dockerfile": "local/builders/java/v2/Dockerfile",
    },
    "depends_on": {
      "service-four-builder": {
        "condition": "service_completed_successfully",
        "restart": true,
      },
    },
    "volumes": [
      "repositories/service-four/target:/opt/",
    ],
  },
  "service-four-builder": {
    "build": {
      "context": "/home/test/user/projects/docker-project",
      "dockerfile": "local/builders/java/v2/build.Dockerfile",
    },
    "volumes": [
      "repositories/service-four:/app/",
      "./out:/opt/out",
    ],
  },
}
`;

exports[`builderSpecAssemblyFunction applies extra properties to service 1`] = `
{
  "service-four": {
    "build": {
      "context": "/home/test/user/projects/docker-project",
      "dockerfile": "local/builders/java/v2/Dockerfile",
    },
    "depends_on": {
      "service-four-builder": {
        "condition": "service_completed_successfully",
        "restart": true,
      },
    },
    "volumes": [
      "repositories/service-four/target:/opt/",
    ],
  },
  "service-four-builder": {
    "build": {
      "context": "/home/test/user/projects/docker-project",
      "dockerfile": "local/builders/java/v2/build.Dockerfile",
    },
    "volumes": [
      "repositories/service-four:/app/",
      "./out:/opt/out",
    ],
  },
}
`;

exports[`builderSpecAssemblyFunction can handle repository builder 1`] = `
{
  "service-four": {
    "build": {
      "context": "/home/test/user/projects/docker-project/repositories/service-four",
    },
  },
}
`;

exports[`builderSpecAssemblyFunction can handle setting dockerfile and context for repo builder 1`] = `
{
  "build": {
    "context": "/home/test/user/projects/docker-project/repositories/service-four/another-dir/",
    "dockerfile": "another.Dockerfile",
  },
}
`;

exports[`builderSpecAssemblyFunction can handle setting dockerfile for repo builder 1`] = `
{
  "build": {
    "context": "/home/test/user/projects/docker-project/repositories/service-three/docker-env",
    "dockerfile": "tilt/Dockerfile",
  },
}
`;

exports[`builderSpecAssemblyFunction can handle setting repository context for repo builder 1`] = `
{
  "build": {
    "context": "/home/test/user/projects/docker-project/repositories/service-eight/tilt/",
  },
}
`;

exports[`builderSpecAssemblyFunction can handle using builder and service dockerfile 1`] = `
{
  "build": {
    "context": "/home/test/user/projects/docker-project/repositories/service-one",
  },
  "depends_on": {
    "service-one-builder": {
      "condition": "service_completed_successfully",
      "restart": true,
    },
  },
  "volumes": [
    "repositories/service-one/target:/opt/",
  ],
}
`;

exports[`builderSpecAssemblyFunction can handle using builder and service dockerfile 2`] = `
{
  "build": {
    "context": "/home/test/user/projects/docker-project",
    "dockerfile": "local/builders/java/v2/build.Dockerfile",
  },
  "volumes": [
    "repositories/service-one:/app/",
    "./out:/opt/out",
  ],
}
`;

exports[`builderSpecAssemblyFunction can handle using builder and service dockerfile appending custom context and dockerfile 1`] = `
{
  "build": {
    "context": "/home/test/user/projects/docker-project/repositories/service-one/another-dir/",
    "dockerfile": "another.Dockerfile",
  },
  "depends_on": {
    "service-one-builder": {
      "condition": "service_completed_successfully",
      "restart": true,
    },
  },
  "volumes": [
    "repositories/service-one/target:/opt/",
  ],
}
`;

exports[`builderSpecAssemblyFunction can handle using builder and service dockerfile appending custom context and dockerfile 2`] = `
{
  "build": {
    "context": "/home/test/user/projects/docker-project",
    "dockerfile": "local/builders/java/v2/build.Dockerfile",
  },
  "volumes": [
    "repositories/service-one:/app/",
    "./out:/opt/out",
  ],
}
`;
